corfudb-tools/src/main/java/org/corfudb/browser/CorfuOfflineBrowserEditor.java

```diff
@@ -0,0 +1,471 @@
+                            }
+                        }
+                    }
+                }
+                System.out.println("Finished processing file: " + file.getAbsolutePath());
+
+            } catch (IOException e) {
+                throw new IllegalStateException("Invalid header: " + file.getAbsolutePath(), e);
+            }
+        }
+    }
+
+    /**
+     * Processes the LogData, deserialize and extracts SMR entries.
+     * @param data LogData.
+     * @param table the Table Descriptor.
+     * @param runtimeSerializer runtime serializer to deserialize logData.
+     * @param tableEntries SMR and Checkpoint updates.
+     */
+    private void processLogData(List<LogData> data,
+                                  CorfuTableDescriptor table,
+                                  CorfuRuntime runtimeSerializer,
+                                  List<LogEntryOrdering> tableEntries) {
+
+        for (int index = 0; index < data.size(); ++index) {
+
+            Object modifiedData = data.get(index).getPayload(runtimeSerializer);
+            List<SMREntry> smrUpdates = new ArrayList<>();
+            long snapshotAddress = 0;
+            if (modifiedData instanceof CheckpointEntry) {
+                snapshotAddress = Long.decode(((CheckpointEntry)modifiedData).
+                        getDict().
+                        get(CheckpointEntry.CheckpointDictKey.SNAPSHOT_ADDRESS));
+                MultiSMREntry smrEntries = ((CheckpointEntry) modifiedData).
+                        getSmrEntries(false, runtimeSerializer);
+                if (smrEntries != null) {
+                    smrUpdates = smrEntries.getUpdates();
+                }
+            } else if (modifiedData instanceof MultiObjectSMREntry) {
+                smrUpdates = ((MultiObjectSMREntry) modifiedData).getSMRUpdates(table.getStreamID());
+            }
+
+            for (SMREntry smrUpdate : smrUpdates) {
+                LogEntryOrdering entry = (modifiedData instanceof CheckpointEntry) ?
+                        new LogEntryOrdering(smrUpdate, snapshotAddress) :
+                        new LogEntryOrdering(smrUpdate, smrUpdate.getGlobalAddress());
+                tableEntries.add(entry);
+            }
+        }
+    }
+
+    /**
+     * Builds the given table.
+     * @param namespace
+     * @param tableName
+     * @return ConcurrentMap
+     */
+    public ConcurrentMap getTableData(String namespace, String tableName) {
+
+        CorfuTableDescriptor table = new CorfuTableDescriptor(namespace, tableName);
+        List<CorfuTableDescriptor> tableDsc = Arrays.asList(registryTableDsc, protobufDescTableDsc, table);
+
+        List<LogData> registryTableEntries = new ArrayList<>();
+        List<LogData> protobufDescriptorTableEntries = new ArrayList<>();
+        List<LogData> dataEntries = new ArrayList<>();
+        List<List<LogData>> tableData = Arrays.asList(registryTableEntries, protobufDescriptorTableEntries, dataEntries);
```

Comment:
One concern I have is regarding the memory requirement for this list for certain tables which have a huge number of updates but very few actual records. In those cases won't this tableData grow huge because in parseLogs() the data is not applied to the ConcurrentMaps?

Reply:
In general, we have a checkpointing cycle done every 15 minutes, so the expected transactions would be the ones that occurred in the last few minutes.
A desirable situation for the OOM is when the checkpointing fails all over time and we could have a lot of transactions.
As mentioned in the description, the next iteration of the tool is to make all the in-memory data structs indexed onto the file system, in that way neither the lists nor the ConcurrentMaps would ever trigger OOM.

Reply:
If so can we modify this PR to not break in the case of checkpointing failures and re-introduce the above optimization once the disk based indexing PR is merged? That way the current version of the tool will not be broken and once the disk based indexing is merged it can take advantage of that?

Also not sure if the disk based indexing will help here where the stream address space for a particular stream itself is so large that it cannot fit in memory (Metadata itself is larger than memory Xmx)

Reply:
I get your point. I have updated the logic to not cache all the transactions, (but with an overhead of parsing all the logs files twice).